timstamp modified

sak797 opened the pull request:
ros::Time( double(pFrameData->timeStamp) \* 1.e-4); was giving the time since the duo_node started, but the ros time is different. I was not able to get the pointcloud as the published pointcloud is very old. The above change worked. I was getting pointcloud. I am doubtful about the type of **leftImage.header.stamp** .

Commit:
timstamp modified

ros::Time( double(pFrameData->timeStamp) * 1.e-4); was giving the time since the duo_node started, but the ros time is different. I was not able to get the pointcloud as the published pointcloud is very old. The above change worked. I was getting pointcloud. I am doubtful about the type of **leftImage.header.stamp** .

bsubei commented:
I'm pretty sure header.stamp is not of type double:
http://docs.ros.org/indigo/api/std_msgs/html/msg/Header.html

It's of type time, probably this one here:
http://docs.ros.org/diamondback/api/rostime/html/classros_1_1Time.html

bsubei commented:
according to the [DUO API](http://duo3d.com/docs/articles/api#PDUOFrameLink), the timestamp is a uint32 given in increments of 100 microseconds. So multiplying them by 10^-4 makes them into seconds. Then ros::Time() takes seconds as an argument. It looks correct to me.

bsubei commented:
Also, the point of using the timestamp from the DUO camera itself (instead of using ros::Time::now) is that it's more accurate (not sure if that's needed, though).

If you want, I guess you could just do:
`leftImage.header.stamp = ros::Time::now();`

sak797 commented:
ok, so that line has to be changed to **ros::Time( double(ros::Time::now().toSec()) )** instead of **double(ros::Time::now().toSec())** , also I am suggesting that the timestamp given by the frame should not be used as it starts only since the node is activated.  ros time is the system time or the time since the epoch(I think).

or some global variable has to be initialized when the duo node is started and that constant has to be added to the timestamp given by the frame.

Also should I be making a new pull request ? I don't how to modify my previous change

bsubei commented:
ok, sorry I previously missed your point about the timestamp being since start. Yes, your suggested fix (variable to store start time) sounds right.

For a quick fix (like your code attempts), just do:
`leftImage.header.stamp = ros::Time::now();`

You can modify your existing pull request by pushing more commits to it (or preferably overwriting the existing one). I recommend doing a `git commit --amend` and then force pushing to your fork (overwriting the existing commit).

bsubei commented:
or you could squash commits. There's [tons](http://stackoverflow.com/questions/9790448/how-to-update-a-pull-request) of hints online :)

Commit:
modified time stamp

I did the change but I dont know if [this](https://github.com/l0g1x/DUO-Camera-ROS/issues/16) problem will arise. **stereo_image_proc** was giving me good disparity maps even after making the change.

sak797 commented:
I did the change, but I dont know if [this](https://github.com/l0g1x/DUO-Camera-ROS/issues/16) problem will arise. **stereo_image_proc** was giving me good disparity maps even after making the change.

bsubei commented:
That looks good. Not sure @l0g1x would want to use ros::Time::now, though, because you don't get the real time when the frame was actually captured by the camera hardware (which is what pFrameData->timeStamp gives, I assume). I'm sure he can implement your suggestion (to make it the same as ros time and not from beginning of node startup), unless you do it first :)

sak797 commented:
I tried the changes , but I keep getting **undefined reference globalvar**. I think it is due my lesser understanding of namespaces. I tried few changes suggested on how to declare global variables but I keep getting this error.

What I did was:
driverDUOstereo.h
extern double starttime;  // declared the global variable after the include statements outside the     namespace(tried inside the namspace but getting **error:storage class specified for 'starttime'**)

node.cpp
double starttime = ros::Time::now().toSec();  //before the **ros::Spin()** and after the **duoDriver.startDUO();**

driverDUOstereo.cpp
ros::Time(g_startTime + (double(pFrameData->timeStamp)_1.e-4 ) );// replaced *__ros::Time::now()_

l0g1x commented:
dont make it extern. just put it under the private variables in the header file. move the `starttime = ros::Time::now().toSec();` into the startDUO() function, and dont redefine it since you already prototyped it in the class header. Im not saying this is the solution, but this should allow you to compile

bsubei commented:
@kanirudh54 I believe this^ is what you're trying to do. No need for a global var, just make a private variable for starttime.

Please try running it (I don't have a DUO right now), but it compiles fine :)

sak797 commented:
Yes that's what I tried but I used **double** instead of **int**.It compiles and works(point cloud on rviz and disparity maps using stereo_image_proc) . I will try it with **int** and let you know.

bsubei commented:
actually, I just realized [toSecs()](http://docs.ros.org/api/rostime/html/time_8h_source.html#l00152) returns double, so you're right about that.

sak797 commented:
okay

l0g1x commented:
So im a bit confused as to what the issue your saying actually is.. For me, i can get a pointcloud just fine (refer to image) without making any changes to the time stamps. I would need to ask some people this question who may know better, but i believe sending the time in which the camera's hardware captured the image is the time that should be used. Are you getting a error when trying to use the tf or tf2 library functions like lookupTransform? if you are then it would make a bit more sense as to why you are having a issue. Im viewing the data in the `duo3d_camera` frame in rviz.

steps i did: 

```
roslaunch duo3d_ros duo.launch

ROS_NAMESPACE=duo3d_camera rosrun stereo_image_proc stereo_image_proc

rosrun image_view stereo_view stereo:=duo3d_camera image:=image_raw

rosrun rviz rviz
```

and then change the global frame in rviz to duo3d_camera

sak797 commented:
I am also using tf library. I am trying to run viso2_ros and fovis. Though I am not getting the desired results, I don't get errors like the transform is missing or messages are dropped.(I think I used to get them due to the difference in timestamps). Did you try any visual odometry algorithms?

l0g1x commented:
okay im going to post a question on ros answers about this as i dont know truely what should be done as I used different other camera drivers as a reference for the data timestamp. I will hopefully have your answer by tomorrow. 

As for trying a visual odometry algorithm, i have not. But after looking at how viso2 processes frame data, i think you would need to use ros::time::now(). Not 100% sure though.

sak797 commented:
okay, thanks

gitunit commented:
so far i tried viso2_ros, fovis and rtabmap. neither of them worked for me. could be very much related to the timestamp. not sure though.
